refactor/ui_backend/app/api/api_v0/endpoints/ui.py

```diff
@@ -17,18 +17,40 @@
 # FITNESS FOR A PARTICULAR PURPOSE. See the License for more details.
 #
 # Relief from the License may be granted by purchasing a commercial license.
-
+import logging 
+from aiolimiter import AsyncLimiter
 
 from app.schemas.ui import UpdateUIResponse
-from fastapi import APIRouter
+from fastapi import APIRouter, Request, BackgroundTasks
+
+dispatch_set = set()
+limiter = AsyncLimiter(1, 2)
 
 router = APIRouter()
 
+async def notify_frontend(app, dispatch_id, task_id):
+    logging.debug(f"Emitting websocket event to update task {task_id} in workflow {dispatch_id}")
+    dispatch_set.remove((dispatch_id, task_id))
+    await app.sio.emit("result-update",{
+        "result": {
+            "dispatch_id": dispatch_id,
+            "task_id": task_id
+        }
+    })
+
+async def add_to_bucket(app, dispatch_id, task_id):
+    global dispatch_set
+    if (dispatch_id, task_id) not in dispatch_set:
+        dispatch_set.add((dispatch_id,task_id))
+        async with limiter:
+            await notify_frontend(app, dispatch_id, task_id)
+        
 
 @router.put("/workflow/{dispatch_id}/task/{task_id}", status_code=200, response_model=UpdateUIResponse)
-def update_ui(*, dispatch_id: str, task_id: int) -> UpdateUIResponse:
+async def update_ui(*, dispatch_id: str, task_id: int, request: Request, background_tasks: BackgroundTasks) -> UpdateUIResponse:
```

Comment:
are `request` and `background_tasks` meant to be query params? if so the swagger spec will need to be updated https://app.swaggerhub.com/apis/agnostiq/Covalent_UI_Service_API/0.1.0#/UI/update_ui_api_v0_ui_workflow__dispatch_id__task__task_id__put

Reply:
@scottwn `request` and `background_tasks` are fastapi dependencies that are injected into all route handlers. `request` is able to get the application context with `request.app` (to avoid circular dependencies), and `background_tasks` is a way to run async coroutines in the background. Hence we do not need to update the api spec as a result.